lib/tests/streamlit/element_utils_test.py

```diff
@@ -0,0 +1,93 @@
+# Copyright 2018-2021 Streamlit Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import unittest
+from unittest.mock import MagicMock, patch
+
+import pytest
+
+from streamlit.elements.utils import check_callback_rules, check_session_state_rules
+from streamlit.errors import StreamlitAPIException
+
+
+class ElementUtilsTest(unittest.TestCase):
+    @patch("streamlit.elements.utils.is_in_form", return_value=False)
+    @patch("streamlit._is_running_with_streamlit", new=True)
+    def test_check_callback_rules_not_in_form(self, _):
+        check_callback_rules(None, lambda x: x)
+
+    @patch("streamlit.elements.utils.is_in_form", return_value=True)
+    @patch("streamlit._is_running_with_streamlit", new=True)
+    def test_check_callback_rules_in_form(self, _):
+        check_callback_rules(None, None)
+
+    @patch("streamlit.elements.utils.is_in_form", return_value=True)
+    @patch("streamlit._is_running_with_streamlit", new=True)
+    def test_check_callback_rules_error(self, _):
+        with pytest.raises(StreamlitAPIException) as e:
+            check_callback_rules(None, lambda x: x)
+
+        assert (
+            str(e.value)
+            == "Callbacks are not allowed on widgets in forms; put them on the form submit button instead."
```

Comment:
This kind of test feels overly brittle, by being coupled to the exact text we display despite only caring about the text as an indicator that we triggered a particular exception case.

Sometime in the future we should consider making our api exceptions more fine-grained, so that the exception type itself conveys this information and the code doesn't have to care about the specific string.

Reply:
Ack -- I tweaked this (and other direct string equality asserts in this test) to something like
```python
assert "Callbacks are not allowed on widgets in forms" in str(e.value)
```

so we still have some validation of the error message without it feeling as brittle as an exact match.

Agreed on making the exceptions more fine-grained in the future so the error message checks are totally unecessary